monai/transforms/spatial/dictionary.py

```diff
@@ -2194,6 +2203,164 @@ def __call__(self, data: Mapping[Hashable, NdarrayOrTensor]) -> List[Dict[Hashab
+            Defaults to image size modulo patch size.
+        fix_num_patches: number of patches to return. Defaults to None, which returns all the available patches.
+        overlap: the amount of overlap of neighboring patches in each dimension (a value between 0.0 and 1.0).
+            If only one float number is given, it will be applied to all dimensions. Defaults to 0.0.
+        sort_fn: a callable or string that defines the order of the patches to be returned. If it is a callable, it
+            will be passed directly to the `key` argument of `sorted` function. The string can be "min" or "max",
+            which are, respectively, the minimum and maximum of the sum of intensities of a patch across all dimensions
+            and channels. Also "random" creates a random order of patches.
+            By default no sorting is being done and patches are returned in a row-major order.
+        pad_mode: {``"constant"``, ``"edge"``, ``"linear_ramp"``, ``"maximum"``, ``"mean"``,
+            ``"median"``, ``"minimum"``, ``"reflect"``, ``"symmetric"``, ``"wrap"``, ``"empty"``}
+            One of the listed string values or a user supplied function. Defaults to ``"wrap"``.
+            See also: https://numpy.org/doc/stable/reference/generated/numpy.pad.html
+        pad_opts: padding options, see `numpy.pad`
+        seed: random seed to generate offsets
+        allow_missing_keys: don't raise exception if key is missing.
+
+
+    """
+
+    backend = RandGridPatch.backend
+
+    def __init__(
+        self,
+        keys: KeysCollection,
+        patch_size: Sequence[int],
+        min_start_pos: Optional[Union[Sequence[int], int]] = None,
+        max_start_pos: Optional[Union[Sequence[int], int]] = None,
+        fix_num_patches: Optional[int] = None,
+        overlap: float = 0.0,
+        sort_fn: Optional[Union[Callable, str]] = None,
+        pad_mode: Union[NumpyPadMode, PytorchPadMode, str] = NumpyPadMode.CONSTANT,
+        pad_opts: Optional[Dict] = None,
+        allow_missing_keys: bool = False,
+    ):
+        MapTransform.__init__(self, keys, allow_missing_keys)
+        self.patcher = RandGridPatch(
+            patch_size=patch_size,
+            min_start_pos=min_start_pos,
+            max_start_pos=max_start_pos,
+            fix_num_patches=fix_num_patches,
+            overlap=overlap,
+            sort_fn=sort_fn,
+            pad_mode=pad_mode,
+            pad_opts=pad_opts,
+        )
+
+    def set_random_state(
+        self, seed: Optional[int] = None, state: Optional[np.random.RandomState] = None
+    ) -> "RandGridPatchd":
+        self.patcher.set_random_state(seed, state)
+        super().set_random_state(seed, state)
+        return self
+
+    def __call__(self, data: Mapping[Hashable, NdarrayOrTensor]) -> List[Dict]:
+        d = dict(data)
+        original_spatial_shape = d[first(self.keys)].shape[1:]
+        output = []
+        results = [self.patcher(d[key]) for key in self.keys]
```

Comment:
Here the `self.patcher` will execute different random behavior for different keys.
If you want to ensure the consistent random behavior for keys, please refer to:
(1) Enable random operation for the dict transform:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/transforms/intensity/dictionary.py#L194
(2) Then disable random operation for the array transform with `randomize=False`:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/transforms/intensity/dictionary.py#L205

Thanks.

Reply:
Thanks, that's an important point. I solved it differently. I believe randomness should be controlled only with random state. There is no point for this transform with `randomize=False` but if we reset the random state in call for each of the keys, then we end up with the same randomization for all of them. Please take a look at the code and let me know what you think,=.

Reply:
Hi @drbeh ,

Sorry I pasted a wrong link for your reference, this is a typical example for the **consistent random behavior of keys**:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/transforms/intensity/dictionary.py#L203-L205
@rijobro @ericspod What do you think?

Thanks.

Reply:
Thanks @Nic-Ma! I see but I believe in this case that randomization does not depend on the value of images, it's more straightforward to manage the randomization with setting `set_random_state` instead of having an additional flag to skip it. Does it make sense?

Reply:
Thanks for the quick update, looks good to me.
Please feel free to raise a discussion with @ericspod @rijobro @wyli about your previous random implementation if necessary.

Thanks.